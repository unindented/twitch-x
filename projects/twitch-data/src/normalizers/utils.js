export function valueOrNull (value) {
  return value || null
}
