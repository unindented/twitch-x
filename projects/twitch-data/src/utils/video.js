import {stringify} from 'qs'
import {USHER_API_URL} from '../constants'

export function getPlaylistUrl ({stream, sig, token}) {
  const query = {
    sig,
    token,
    allow_audio_only: true,
    allow_source: true,
    player: 'twitchweb',
    type: 'any',
    p: Math.floor(Math.random() * 1e7)
  }
  return `${USHER_API_URL}/channel/hls/${stream}.m3u8?${stringify(query)}`
}

export function extractFeedsFromPlaylist (input) {
  const regexp = /#EXT-X-MEDIA:.*NAME="([^"]+)".*\n#EXT-X-STREAM-INF:.*\n(.*)/g
  const result = []

  let match = regexp.exec(input)
  while (match != null) {
    result.push({
      name: match[1],
      url: match[2]
    })
    match = regexp.exec(input)
  }

  return result
}
