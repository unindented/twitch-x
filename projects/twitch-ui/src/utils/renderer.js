import React from 'react'
import renderer from 'react-test-renderer'
import {ThemeProvider} from 'styled-components'
import themes from '../themes'

export function renderWithTheme (node, theme = 'default') {
  return renderer.create(
    <ThemeProvider theme={themes[theme]}>
      {node}
    </ThemeProvider>
  )
}
