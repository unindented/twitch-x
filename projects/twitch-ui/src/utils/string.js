export function supplant (str, obj) {
  return str.replace(/\{\s*([^{}]+?)\s*\}/g, (str, key) => obj[key])
}
