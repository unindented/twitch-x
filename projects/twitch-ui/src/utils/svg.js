export function rectangle ({width, height, fill}) {
  return `<svg viewBox="0 0 ${width} ${height}" xmlns="http://www.w3.org/2000/svg"><rect fill="${fill}" width="100%" height="100%" /></svg>`
}
