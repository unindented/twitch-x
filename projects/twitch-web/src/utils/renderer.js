import ShallowRenderer from 'react-test-renderer/shallow'

export function render (node, theme = 'default') {
  const renderer = new ShallowRenderer()
  renderer.render(node)
  return renderer.getRenderOutput()
}
