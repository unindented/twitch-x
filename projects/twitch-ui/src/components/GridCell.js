import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import ImageWithPlaceholder from './ImageWithPlaceholder'
import {truncate} from '../utils/css'

const GridCellRoot = styled.a`
  display: block;
  overflow: hidden;
  flex: 1;
  padding: ${props => props.theme.layout.gaps.medium};
  color: inherit;
  text-decoration: none;

  &:hover {
    color: ${props => props.theme.colors.highlight};
  }
`

const GridCellFrame = styled.div`
  padding: ${props => props.theme.layout.borders.small};

  /* stylelint-disable-next-line */
  ${GridCellRoot}:hover & {
    border: ${props => props.theme.layout.borders.medium} solid;
    margin: -${props => props.theme.layout.borders.medium};
  }
`

const GridCellName = styled.h3`
  margin: ${props => props.theme.layout.gaps.small} 0 0;
  font-size: ${props => props.theme.font.sizes.secondary};
  font-weight: normal;
  ${truncate}
`

const GridCellInfo = styled.p`
  margin: 0;
  font-size: ${props => props.theme.font.sizes.tertiary};
  ${truncate}
`

export default function GridCell (props) {
  const {href, image, name, viewers, channels, width, height, focused} = props

  return (
    <GridCellRoot
      href={href}
    >
      <GridCellFrame
        focused={focused}
      >
        <ImageWithPlaceholder
          src={image}
          width={width}
          height={height}
        />
      </GridCellFrame>
      <GridCellName>
        {name}
      </GridCellName>
      {viewers && (
        <GridCellInfo>
          {viewers} viewers
        </GridCellInfo>
      )}
      {channels && (
        <GridCellInfo>
          {channels} channels
        </GridCellInfo>
      )}
    </GridCellRoot>
  )
}

GridCell.propTypes = {
  href: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  viewers: PropTypes.number,
  channels: PropTypes.number,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  focused: PropTypes.bool
}

GridCell.defaultProps = {
  focused: false
}
