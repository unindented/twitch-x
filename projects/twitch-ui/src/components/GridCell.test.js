import React from 'react'
import {renderWithTheme} from '../utils/renderer'
import GridCell from './GridCell'

describe('GridCell', () => {
  let tree

  beforeEach(() => {
    tree = renderWithTheme(
      <GridCell
        href='/games/488552'
        image='//lorempixel.com/{width}/{height}/'
        name='Overwatch'
        viewers={41343}
        width={272}
        height={380}
      />
    )
  })

  it('renders', () => {
    expect(tree).toMatchSnapshot()
  })
})
