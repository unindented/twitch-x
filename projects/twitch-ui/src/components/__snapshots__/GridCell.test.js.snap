// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCell renders 1`] = `
.c3 {
  display: block;
  width: 100%;
  height: auto;
}

.c4 {
  display: none;
  width: 100%;
  height: auto;
}

.c1 {
  display: block;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 10px;
  color: inherit;
  text-decoration: none;
}

.c1:hover {
  color: #6441a4;
}

.c2 {
  padding: 1px;
}

.c0:hover .c2 {
  border: 3px solid;
  margin: -3px;
}

.c5 {
  margin: 5px 0 0;
  font-size: 16px;
  font-weight: normal;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c6 {
  margin: 0;
  font-size: 13px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<a
  className="c0 c1"
  href="/games/488552"
>
  <div
    className="c2"
  >
    <div>
      <img
        alt=""
        className="c3"
        hidden={false}
        src="data:image/svg+xml;utf8,<svg viewBox=\\"0 0 272 380\\" xmlns=\\"http://www.w3.org/2000/svg\\"><rect fill=\\"#6441a4\\" width=\\"100%\\" height=\\"100%\\" /></svg>"
      />
      <img
        alt=""
        className="c4"
        hidden={true}
        onLoad={[Function]}
        src="//lorempixel.com/272/380/"
      />
    </div>
  </div>
  <h3
    className="c5"
  >
    Overwatch
  </h3>
  <p
    className="c6"
  >
    41343
     viewers
  </p>
</a>
`;
